File: akka-samples/akka-sample-remote-scala/src/main/scala/sample/remote/calculator/LookupActor.scala
package sample.remote.calculator

import scala.concurrent.duration._
import akka.actor.Actor
import akka.actor.ActorIdentity
import akka.actor.ActorRef
import akka.actor.Identify
import akka.actor.ReceiveTimeout
import akka.actor.Terminated

class LookupActor(path: String) extends Actor {

  sendIdentifyRequest()

  def sendIdentifyRequest(): Unit = {
    context.actorSelection(path) ! Identify(path)
    import context.dispatcher
    context.system.scheduler.scheduleOnce(3.seconds, self, ReceiveTimeout)
  }

  def receive = identifying

  def identifying: Actor.Receive = {
    case ActorIdentity(`path`, Some(actor)) =>
      context.watch(actor)
      context.become(active(actor))
    case ActorIdentity(`path`, None) => println(s"Remote actor not available: $path")
    case ReceiveTimeout              => sendIdentifyRequest()
    case _                           => println("Not ready yet")
  }

  def active(actor: ActorRef): Actor.Receive = {
    case op: MathOp => actor ! op
    case result: MathResult => result match {
      case AddResult(n1, n2, r) =>
        printf("Add result: %d + %d = %d\n", n1, n2, r)
      case SubtractResult(n1, n2, r) =>
        printf("Sub result: %d - %d = %d\n", n1, n2, r)
    }
    case Terminated(`actor`) =>
      println("Calculator terminated")
      sendIdentifyRequest()
      context.become(identifying)
    case ReceiveTimeout =>
    // ignore

  }
}

File: akka-samples/akka-sample-remote-java/src/main/java/sample/remote/calculator/LookupActor.java
package sample.remote.calculator;

import static java.util.concurrent.TimeUnit.SECONDS;
import scala.concurrent.duration.Duration;
import akka.actor.ActorRef;
import akka.actor.ActorIdentity;
import akka.actor.Identify;
import akka.actor.Terminated;
import akka.actor.UntypedActor;
import akka.actor.ReceiveTimeout;
import akka.japi.Procedure;

public class LookupActor extends UntypedActor {

  private final String path;
  private ActorRef calculator = null;

  public LookupActor(String path) {
    this.path = path;
    sendIdentifyRequest();
  }

  private void sendIdentifyRequest() {
    getContext().actorSelection(path).tell(new Identify(path), getSelf());
    getContext()
        .system()
        .scheduler()
        .scheduleOnce(Duration.create(3, SECONDS), getSelf(),
            ReceiveTimeout.getInstance(), getContext().dispatcher(), getSelf());
  }

  @Override
  public void onReceive(Object message) throws Exception {
    if (message instanceof ActorIdentity) {
      calculator = ((ActorIdentity) message).getRef();
      if (calculator == null) {
        System.out.println("Remote actor not available: " + path);
      } else {
        getContext().watch(calculator);
        getContext().become(active, true);
      }

    } else if (message instanceof ReceiveTimeout) {
      sendIdentifyRequest();

    } else {
      System.out.println("Not ready yet");

    }
  }

  Procedure<Object> active = new Procedure<Object>() {
    @Override
    public void apply(Object message) {
      if (message instanceof Op.MathOp) {
        // send message to server actor
        calculator.tell(message, getSelf());

      } else if (message instanceof Op.AddResult) {
        Op.AddResult result = (Op.AddResult) message;
        System.out.printf("Add result: %d + %d = %d\n", result.getN1(),
            result.getN2(), result.getResult());

      } else if (message instanceof Op.SubtractResult) {
        Op.SubtractResult result = (Op.SubtractResult) message;
        System.out.printf("Sub result: %d - %d = %d\n", result.getN1(),
            result.getN2(), result.getResult());

      } else if (message instanceof Terminated) {
        System.out.println("Calculator terminated");
        sendIdentifyRequest();
        getContext().unbecome();

      } else if (message instanceof ReceiveTimeout) {
        // ignore

      } else {
        unhandled(message);
      }

    }
  };
}
